lib/utils/hfbFileRaw.cpp

```diff
@@ -0,0 +1,221 @@
+    file_metadata["structure"]["metadata_size"] = metadata_size;
+    file_metadata["structure"]["data_size"] = data_size;
+    file_metadata["structure"]["frame_size"] = frame_size;
+    file_metadata["structure"]["nfreq"] = nfreq;
+    file_metadata["structure"]["num_beams"] = num_beams;
+    file_metadata["structure"]["num_subfreq"] = num_subfreq;
+
+    // Create lock file and then open the other files
+    lock_filename = create_lockfile(_name);
+    metadata_file = std::ofstream(_name + ".meta", std::ios::binary);
+    if ((fd = open((_name + ".data").c_str(), oflags,
+                   S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH))
+        == -1) {
+        throw std::runtime_error(
+            fmt::format(fmt("Failed to open file {:s}.data: {:s}."), _name, strerror(errno)));
+    }
+
+    // Preallocate data file (without increasing the length)
+#ifdef __linux__
+    // Note not all versions of linux support this feature, and they don't
+    // include the macro FALLOC_FL_KEEP_SIZE in that case
+#ifdef FALLOC_FL_KEEP_SIZE
+    fallocate(fd, FALLOC_FL_KEEP_SIZE, 0, frame_size * nfreq * max_time);
+#else
+    (void)max_time; // Suppress warning
+    WARN("fallocate not supported on this system!");
+#endif
+#else
+    (void)max_time; // Suppress warning
+#endif
+}
+
+hfbFileRaw::~hfbFileRaw() {
+
+    // Finalize the metadata file
+    file_metadata["structure"]["ntime"] = num_time();
+    file_metadata["index_map"]["time"] = times;
+    std::vector<uint8_t> t = nlohmann::json::to_msgpack(file_metadata);
+    metadata_file.write((const char*)&t[0], t.size());
+    metadata_file.close();
+
+    // TODO: final sync of data file.
+    close(fd);
+
+    std::remove(lock_filename.c_str());
+}
+
+size_t hfbFileRaw::num_time() {
+    return times.size();
+}
+
+void hfbFileRaw::flush_raw_async(int ind) {
+#ifdef __linux__
+    size_t n = nfreq * frame_size;
+    sync_file_range(fd, ind * n, n, SYNC_FILE_RANGE_WRITE);
+#else
+    (void)ind;      // Suppress warning
+#endif
+}
+
+void hfbFileRaw::flush_raw_sync(int ind) {
+#ifdef __linux__
+    size_t n = nfreq * frame_size;
+    sync_file_range(fd, ind * n, n,
+                    SYNC_FILE_RANGE_WAIT_BEFORE | SYNC_FILE_RANGE_WRITE
+                        | SYNC_FILE_RANGE_WAIT_AFTER);
+    posix_fadvise(fd, ind * n, n, POSIX_FADV_DONTNEED);
+#else
+    (void)ind;      // Suppress warning
+#endif
+}
+
+uint32_t hfbFileRaw::extend_time(time_ctype new_time) {
+
+    size_t ntime = num_time();
+
+    // Start to flush out older dataset regions
+    uint delta_async = 2;
+    if (ntime > delta_async) {
+        flush_raw_async(ntime - delta_async);
+    }
+
+    // Flush and clear out any really old parts of the datasets
+    uint delta_sync = 4;
+    if (ntime > delta_sync) {
+        flush_raw_sync(ntime - delta_sync);
+    }
+
+    times.push_back(new_time);
+
+    // Extend the file length for the new time
+#ifdef __linux__
+    fallocate(fd, 0, 0, frame_size * nfreq * num_time());
+#else
```

Comment:
Is there any possibility to merge common functionality of visFileRaw and hfbFileRaw? Either base one on the other, or have a common base class. There's a large amount of the implementation that's completely common.